lightning/src/ln/peers/handshake/states.rs

```diff
@@ -0,0 +1,897 @@
+use bitcoin::secp256k1;
+
+use bitcoin::hashes::{Hash, HashEngine};
+use bitcoin::hashes::sha256::Hash as Sha256;
+use bitcoin::secp256k1::{SecretKey, PublicKey};
+
+use ln::peers::{chacha, hkdf5869rfc};
+use ln::peers::conduit::{Conduit, SymmetricKey};
+use ln::peers::handshake::acts::{Act, ActBuilder, ACT_ONE_LENGTH, ACT_TWO_LENGTH, ACT_THREE_LENGTH, EMPTY_ACT_ONE, EMPTY_ACT_TWO, EMPTY_ACT_THREE};
+
+// Alias type to help differentiate between temporary key and chaining key when passing bytes around
+type ChainingKey = [u8; 32];
+
+// Generate a SHA-256 hash from one or more elements concatenated together
```

Comment:
> Initial @arik-so  review: I still think this macro might be more legibly put in a function, or a custom hash object/class. But, as I said before, ultimately this is up to Matt and Antoine.

Overall, I think it makes the code less readable and introduces a wrapper type that has to be unwrapped which I think detracts from the audit-ability. I'm also not sure about the aversion to variadic macros. I needed a way to do the hash for 1 or 2 objects and used a macro that made the code (in my point of view) more legible.

With that said, I did implement it to get your take and see which way the code should go. Just let me know and I can fixup this PR to have the final result.

julianknutsen@0b4c819